### ISSUE TITLE
Extend template to accept namespace (pid)

### ISSUE THREAD

**Issue by gianarb** [tinkerbell/tink#434]
Hello

This morning I had some fun with @thebsdbox pairing on writing action to kexec inside a new shiny operating system.

The outcome is this action https://github.com/tinkerbell/hub/pull/8 . It almost works. What is missed is the ability to attach the PID namespace of the host to the action itself.

We even streamed out work on [Twitch](https://www.twitch.tv/videos/900468397) and as you can see at the end of the video we validated that running the action locally with `docker run --pid=host` made the trick and kexec successfully.

So we are almost there but we need to extend the template in a way that will allow passing the pid namespace. 

There are at least two possible solutions.

So we can add the field `action.pid=string` to Template in this way we can pass that value to the Docker API.

PRO: It is simple
BUT I prefer to go a little bit deeper and see what the [OCI](https://github.com/opencontainers/runtime-spec) has to offer because ideally, I would like to be able to replace Docker with other things in the future and I am wondering `pid=<string>` will make it harder. If OCI sounds too low level at least having a look at contained...

**Comment by thebsdbox** [comment]
Looking at the spec, we can later map the pid=<string> to either a named namespace or check if it's an int and map to the namespace ID.  https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#namespaces

I'm in favour of the simple option, it matches the underlying runtime Docker .. which matches its underlying runtime containerd.

**Comment by gianarb** [comment]
@thebsdbox yep, I know that docker relays on what containerd can do that relays on what runc can do, the question is how much it is hidden in those layers and what do we want to support?

```
--pid=""  : Set the PID (Process) Namespace mode for the container,
             'container:<name|id>': joins another container's PID namespace
             'host': use the host's PID namespace inside the container
```

Do we want to act as a proxy and accept all the possible parameters? How it will be even possible or useful to use `pid: container:name`? Do we want to share PID namespace between actions (so mapping container:name with action:name) maybe? 

I know @jacobweinstock has an idea and @markyjackson-taulia will come up with a possible implementation. I personally don't have an idea about how it should be implemented but I would like to a clear and hard to get wrong feature.

**Comment by jacobweinstock** [comment]
I lean toward using something like `pid: host` or `hostPid: true` for now, as we know the use cases (kexec, reboot) we need it for, and then revisit accepting others in the future when other use cases arise. 

I think there are a few schools of thought here found in other projects.
[docker](https://github.com/tinkerbell/tink/issues/434#issuecomment-773568445): `see Gianluca's comment above`
[linuxkit yaml](https://github.com/linuxkit/linuxkit/blob/master/docs/yaml.md): `pid sets the pid namespace, either to a path, or if host is specified it will use the host namespace.`
[kubernetes pod spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#pod-v1-core): `hostPID: true - Use the host's pid namespace. Optional: Default to false.`
[kubernetes sharing pid namespace between containers in a pod](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/): `shareProcessNamespace: true`
[containerd](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#namespaces): `uses the OCI spec which Dan commented on above`

**Comment by thebsdbox** [comment]
> I lean toward using something like `pid: host`

+1

**Comment by gianarb** [comment]
> I lean toward using something like pid: host

@thebsdbox @jacobweinstock I think with `pid: string` you are suggesting an enum? Or just an empty string? What other value should it get when it is not `host`?

**Comment by thebsdbox** [comment]
This is a bit of a blocker on new actions we're wanting to write, has there been the opportunity to write anything so far?

**Comment by markyjackson-taulia** [comment]
@thebsdbox I am starting this wrk today

**Comment by thebsdbox** [comment]
Awesome, I did a PoC today so I could carry on playing -> https://github.com/thebsdbox/tink/tree/pid feel free to pull anything from there that makes sense. I've never used `protos` before so I had to kinda make it up as I was going along 😬  I also missed the db bit so all the bits looked like they should work but the pid stuff didn't exist inside the tink-worker 🙄

**Comment by thebsdbox** [comment]
This can be closed with #436
